### ISSUE TITLE
Missing Darknet neural network support: [shortcut] layers

### ISSUE THREAD

**Issue by VideoPlayerCode** [opencv/opencv#15724]
Ping @dkurt @liuyuns @dmonterom @vpisarev @arthur-williams @alalek @AlexeyAB (everyone who has ever worked on OpenCV's darknet module).

## New Neural Network Innovation!

There's an incredible new neural network innovation: YOLOv3-Tiny has been reduced to almost half CPU usage while retaining detection accuracy. They call it YOLOv3-Tiny-PRN. It's the new king of low-CPU usage networks! Here is the research and paper in progress (soon to be published): https://github.com/WongKinYiu/PartialResidualNetworks

Here's the performance difference on Darknet (it would be even faster in OpenCV):

![66791677-c76bbb00-ef27-11e9-93fe-842dab485bb1](https://user-images.githubusercontent.com/38923130/66974552-b2179d80-f09b-11e9-9a3d-b884e59fb9d6.png)

If you do a diff between yolov3-tiny-prn config and the original yolov3-tiny, you will see that the changes are that many layers have half the filter sizes, and there are many `[shortcut]` layers added instead.

You can also read their paper at the repo above.

The problem is that OpenCV's darknet loader lacks full support for `[shortcut]` layers. So nobody can enjoy this big PRN innovation on OpenCV's ultra-fast CPU neural network implementation...

## References

**Reference: Darknet's implementation of shortcut layers.**

- https://github.com/pjreddie/darknet/blob/b13f67bfdd87434e141af532cdb5dc1b8369aa3b/src/parser.c#L539-L556
- https://github.com/pjreddie/darknet/blob/f6d861736038da22c9eb0739dca84003c5a5e275/src/shortcut_layer.c#L9-L39

**Reference: OpenCV's incomplete implementation of shortcut layers.**

- https://github.com/opencv/opencv/blob/8c0b0714e76efef4a8ca2a7c410c60e55c5e9829/modules/dnn/src/darknet/darknet_io.cpp#L631-L641
- https://github.com/opencv/opencv/blob/8c0b0714e76efef4a8ca2a7c410c60e55c5e9829/modules/dnn/src/darknet/darknet_io.cpp#L412-L432
- We can see one big issue immediately: OpenCV's reading of the shortcut layers only reads the `from=` property. It totally ignores all other properties (mainly the important `activation=` property).
- As a sidenote: The PRN network uses `[shortcut]` with `activation=leaky` (leaky RELU).
- Here's @dkurt 16 months ago, confirming that OpenCV's implementation of `[shortcut]` is wrong (basically it just implements the bare-minimum "Linear" mode for equally-sized layers which are used by the main Yolov3 config, but more advanced usages do not work): https://github.com/opencv/opencv/issues/11568#issuecomment-397277134

**Reference: OpenCV's "Eltwise" layer type, with the "op:sum" parameter, that the YOLO "shortcut" layer is being converted into.**

- https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/eltwise_layer.cpp
- This will most likely need improvement to add a flag which says "allow summing layers of different sizes"... and check how darknet does it; I assume that summing differently shaped arrays is done via interpolated scaling of the source layer's output to the target layer's array size.

**Reference: Official pjreddie configs which use shortcut in various ways.**

- https://github.com/pjreddie/darknet/blob/master/cfg/yolov3.cfg
- All of the ResNet configs at https://github.com/pjreddie/darknet/tree/master/cfg

**Reference: Python script which completely crashes python due to OpenCV failing to read the shortcut layers.**

- cfg: https://github.com/WongKinYiu/PartialResidualNetworks/blob/master/cfg/yolov3-tiny-prn.cfg
- weights: https://github.com/WongKinYiu/PartialResidualNetworks/blob/master/model/yolov3-tiny-prn.weights

```python
import os
import cv2

cfg = "yolov3-tiny-prn.cfg"
weights = "yolov3-tiny-prn.weights"

# True, True
print(os.path.isfile(cfg))
print(os.path.isfile(weights))

# This kills python silently. The second "x" does not execute.
print("x") # prints
foo = cv2.dnn.readNetFromDarknet(cfg, weights) # kills python
print("x")
```

## Summary

- Innovative new Neural Network massively reduces CPU usage and inference time, while keeping the same detection accuracy.
- OpenCV's darknet parser does not support options (such as `activation=leaky`) on the `[shortcut]` layers. The `darknet/parser.c` that I linked above shows that the official list of `[shortcut]` options is `activation=`, `from=`, `alpha=` and `beta=`.
- OpenCV does not support "shortcut"-linking between layers of different sizes, apparently, according to @dkurt.
- OpenCV dies without throwing any descriptive error when trying to load the network's config, as seen in the python file (which shows no error output; it just dies).
- If this can be fixed, OpenCV will be the world's fastest CPU-based framework for running the world's fastest YOLO object detector for low-power devices. Yay. I hope that day is possible in the future...
- Sidenote regarding that last point: With the _current_ official YOLOv3-Tiny in OpenCV, my i7-8750H is taking ~30ms per forward pass (33.3 FPS). If the performance ratio of YOLOv3-Tiny-PRN's speed improvement is the same in OpenCV as it was in Darknet, the new PRN-based network would take ~18.5ms per forward pass (54 FPS) if OpenCV could support it. That's indeed amazing for a CPU detector!

**Comment by dkurt** [comment]
Wow, thank you for such detailed feature request :) You did a lot of work to prepare it which motivates us to help you as fast as possible.

**Comment by VideoPlayerCode** [comment]
@dkurt Thank you! Well, I am trying to help as much as possible with the necessary research since I wasn't confident enough to try to write the changes to OpenCV's Eltwise layer myself! :-)

Very happy to see it targeted for the next release milestone. Thanks @alalek.

If there's anything else I can do to help, just let me know.

**Comment by VideoPlayerCode** [comment]
## Small Correction / Extra Information

The YOLOv3-Tiny-PRN network was designed on @AlexeyAB 's continuation of darknet (which ceased development at the original repo years ago).

Therefore, it's best to use their code as reference instead. Here are the interesting, relevant files I found for analyzing how the shortcut layer works:

- https://github.com/AlexeyAB/darknet/blob/master/src/shortcut_layer.c
- https://github.com/AlexeyAB/darknet/blob/master/src/shortcut_layer.h
- https://github.com/AlexeyAB/darknet/blob/master/src/parser.c (contains the `parse_shortcut` function which reads the config file and turns it into a Shortcut layer)
- https://github.com/AlexeyAB/darknet/blob/master/src/parser.h
- https://github.com/AlexeyAB/darknet/commit/1c71f001531a5df0637903117c6568725d7a66b3 (A commit that I don't understand... it's named "Fixed shortcut_layer for antialiasing", perhaps @AlexeyAB can clarify. But from reading, it seems like it's saying "If the from-layer that we're copying data from is marked as 'antialiased', make an antialiased copy.")

And perhaps I should also tag @WongKinYiu in this discussion, since the goal of their YOLOv3-Tiny-PRN network was fast inference on low-power devices (including GPU-less devices), and OpenCV is the _best_ CPU-based engine. :-)

**Comment by WongKinYiu** [comment]
@VideoPlayerCode Hello,

Both of official darknet and AlexeyAB's darknet can work with PRN.

And the anti-aliasing is for the issue [Antialiased CNNs (ICML 2019)](https://github.com/AlexeyAB/darknet/issues/3672).

@AlexeyAB gives huge contributions for bringing soooo many new functions into darknet.
[The repo](https://github.com/AlexeyAB/darknet) supports many state-of-the-art methods.
eg. [TridentNet (ICCV 2019)](https://github.com/AlexeyAB/darknet/issues/3363), [EfficientNet (ICML 2019)](https://github.com/AlexeyAB/darknet/issues/3380), and [GIoU (CVPR 2019)](https://github.com/AlexeyAB/darknet/issues/3249)...

Our goal is design networks for witch works on low-power ASIC such as NPU or TPU.
Of course the design also benefit on CPU.
For CPU, I suggest using depth-wise-convolution-based architectures, for example MobileNet-v3 (ICCV 2019), ShuffleNet-v2 (ECCV 2018), and EfficientNet (ICML 2019).
Also, we  will release our developed MobileNet-v2-based model in near future.

**Comment by VideoPlayerCode** [comment]
@WongKinYiu Hello again! Thanks for the info! It's really great that you knew what the antialiasing was for!

And yeah, Alexey is incredible and has done so much for Darknet and the world of neural networks. He's a genius!

> Our goal is design networks for which works on low-power ASIC such as NPU or TPU.
> Of course the design also benefit on CPU.

Yeah as your chart showed, you managed to reduce **CPU** inference time of YOLOv3-Tiny to 62.5% in your PRN variant, which is almost half twice as fast as standard! That's incredible.

> For CPU, I suggest using depth-wise-convolution-based architectures with tvm, for example MobileNet-v3 (ICCV 2019), ShuffleNet-v2 (ECCV 2018), and EfficientNet (ICML 2019).

The only one I hadn't heard / tried there is ShuffleNet.

As for **MobileNet**, its detector is meant for mobile phones. Its accuracy is aaaaawful, so I've always avoided it. Look at its mAP at COCO 0.5, where it gets mAP scores of 22 and 16:

![mobilenet](https://user-images.githubusercontent.com/38923130/67059617-100aba80-f15a-11e9-808d-d952a1b253d5.jpg)

Now compare that to your **YOLOv3-Tiny-PRN** at COCO 0.5, whose mAP is 33.1:

![yolo3tinyprn](https://user-images.githubusercontent.com/38923130/66974552-b2179d80-f09b-11e9-9a3d-b884e59fb9d6.png)

About **EfficientNet**, it's indeed a very efficient network and I was super impressed when I read about how they had developed a technique for optimizing and properly scaling the network to halve CPU usage while keeping or even _increasing_ accuracy, thanks to the improved net design. So yeah that's great, but it's slower than YOLOv3-Tiny, and EfficientNet is too new right now so it's lacking good Python libraries and variants (there are some small libraries but nothing official/standard yet). I'm waiting for a Tensorflow2 rewrite of EfficientNet (read that Google is working on rewriting all nets for that, slooowly), and will give it another chance then. It was also designed for Image Classification, _not_ Object Detection, although I am aware that some people have made tweaks for use in object detection, but yet again there's nothing official yet. I saw some Google engineers say that they'll write an object detector version on its official tensorflow pages, but nothing has come from that yet.

I'll research ShuffleNet which you mentioned. A quick search found https://arxiv.org/abs/1807.11164 which seems like a good network, and I'll be reading the whole and testing it.

However, the thing about YOLOv3-Tiny (and now YOLOv3-Tiny-PRN) is that they offer _high_ mAP while being super fast on CPUs. And your PRN design innovation helped a lot on CPU too (on the Darknet engine). So right now YOLOv3-Tiny-PRN is the # 1 most interesting CPU network for me, since you've managed to keep the accuracy of the already proven YOLOv3-Tiny, but made it almost twice as fast on CPU thanks to your redesign! _That_ is what excites me!

> Also, we will release our developed MobileNet-v2-based model in near future.

I look forward to exploring that. You're doing great work!

**Comment by dkurt** [comment]
@VideoPlayerCode, Hi! Can you please the changes from https://github.com/opencv/opencv/pull/15739 with the model?

**Comment by VideoPlayerCode** [comment]
This has now been merged. For anyone who wants to read the discussion, or just jump to some performance results (comparing **YOLOv3-Tiny** vs **YOLOv3-Tiny-PRN**), check this post:

https://github.com/opencv/opencv/pull/15739#issuecomment-544931445

Awesome neural network, and awesome OpenCV engine! <3
